pymatviz/_preprocess_data.py

```diff
@@ -0,0 +1,264 @@
+            to preprocess.
+
+    Returns:
+        pd.DataFrame: The preprocessed DataFrame with element names
+            as index and values as columns.
+
+    Example:
+        >>> data_dict: dict = {
+            "H": 1.0,
+            "He": [2.0, 4.0],
+            "Li": [[6.0, 8.0], [10.0, 12.0]],
+        }
+
+        OR
+        >>> data_df: pd.DataFrame = pd.DataFrame(
+            data_dict.items(),
+            columns=["element", "heat_val"]
+            ).set_index("element")
+
+        OR
+        >>> data_series: pd.Series = pd.Series(data_dict)
+
+        >>> preprocess_data(data_dict / df / series)
+
+        >>> data_df
+            Element   Value
+            H         [1.0, ]
+            He        [2.0, 4.0]
+            Li        [[6.0, 8.0], [10.0, 11.0]]
+
+        Metadata (data_df.attrs):
+            vmin: 1.0
+            mean: 6.0
+            vmax: 11.0
+    """
+
+    def format_pd_dataframe(df: pd.DataFrame) -> pd.DataFrame:
+        """Format pd.DataFrame that may not meet expected format."""
+
+        def fix_df_elem_as_col(df: pd.DataFrame) -> pd.DataFrame:
+            """Fix pd.DataFrame where elements are in a single column."""
+            # Copy and reset index to move element names to a column
+            new_df = df.copy().reset_index()
+
+            # Find the column with element names
+            elem_col_name = None
+            for col in new_df.columns:
+                if set(new_df[col]).issubset({str, Element}):
+                    elem_col_name = col
+                    break
+
+            # Fix failed: cannot find the elements column
+            if elem_col_name is None:
+                return None
+
+            # Rename the column with elements and set it as index
+            new_df = new_df.rename(columns={elem_col_name: Key.element})
+            new_df = new_df.set_index(Key.element)
+
+            # Zip the remaining values into a single column
+            value_cols = [
+                col_name for col_name in new_df.columns if col_name != Key.element
+            ]
+            new_df[Key.heat_val] = new_df[value_cols].apply(list, axis="columns")
+
+            # Drop the old value columns
+            return new_df[[Key.heat_val]]
+
+        # Check if input DataFrame is in expected format
+        if Key.element == df.index.name and Key.heat_val in df.columns:
+            return df
+
+        # Re-format it to expected
+        warnings.warn("DataFrame has unexpected format, trying to fix.", stacklevel=2)
+
+        # Try to search for elements as a column
+        fixed_df = fix_df_elem_as_col(df)
+        if fixed_df is not None:
+            return fixed_df
+
+        # Try to search for elements as a row
+        fixed_df = fix_df_elem_as_col(df.transpose())
+        if fixed_df is not None:
+            return fixed_df
+
+        raise ValueError(f"Cannot normalize the given DataFrame:\n{df.describe()}")
```

Comment:
can we add a test case that triggers this error msg?

Reply:
Done. Basically any `dataframe` that doesn't have a complete row/column of elements should trigger this.